File: api-node/src/middleware/errorHandler.js
/**
 * Centralized error handling — maps AppError and unknown errors to JSON.
 */
import { AppError, toErrorResponse } from '../utils/errors.js';
import { logger } from '../utils/logger.js';

export function notFoundHandler(req, res) {
  res.status(404).json({
    success: false,
    error: { code: 'NOT_FOUND', message: `route ${req.method} ${req.path} not found` },
  });
}

export function errorHandler(err, req, res, next) {
  if (res.headersSent) {
    return next(err);
  }

  const status = err instanceof AppError ? err.statusCode : 500;
  if (status >= 500) {
    logger.error({ err }, 'unhandled error');
  }

  res.status(status).json(toErrorResponse(err));
}

File: api-go/internal/middleware/error_handler.go
package middleware

import (
	"errors"

	"github.com/gofiber/fiber/v2"
	"github.com/interseguros/challenge/api-go/internal/utils"
)

// ErrorHandler maps AppError and unknown errors to consistent JSON responses.
func ErrorHandler(c *fiber.Ctx, err error) error {
	code := fiber.StatusInternalServerError
	var appErr *utils.AppError

	if errors.As(err, &appErr) {
		code = appErr.HTTPStatus
		return c.Status(code).JSON(utils.ErrorResponse{
			Success: false,
			Error:   appErr,
		})
	}

	// Fiber body parser and similar may return *fiber.Error
	var fiberErr *fiber.Error
	if errors.As(err, &fiberErr) {
		code = fiberErr.Code
	}

	return c.Status(code).JSON(utils.ErrorResponse{
		Success: false,
		Error: &utils.AppError{
			Code:       "INTERNAL_ERROR",
			Message:    err.Error(),
			HTTPStatus: code,
		},
	})
}
